lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Iterators;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.Slices;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ChunkedInputStream
+        extends InputStream
+{
+    private final Iterator<SliceInput> chunks;
+    private SliceInput current;
+
+    public static ChunkedInputStream from(Iterator<Slice> slices)
+    {
+        return new ChunkedInputStream(Iterators.transform(slices, Slice::getInput));
+    }
+
+    @VisibleForTesting
+    ChunkedInputStream(Iterator<SliceInput> chunks)
+    {
+        this.chunks = requireNonNull(chunks, "chunks is null");
+        current = chunks.hasNext() ? chunks.next() : Slices.EMPTY_SLICE.getInput();
+    }
+
+    public Slice getSlice(int length)
+    {
+        if (length == 0) {
+            return Slices.EMPTY_SLICE;
+        }
+        while (current.available() == 0) {
+            checkArgument(chunks.hasNext(), "requested %s bytes but 0 was available", length);
+            current = chunks.next();
+        }
+        if (current.available() >= length) {
+            return current.readSlice(length);
+        }
+        // requested length crosses the slice boundary
+        byte[] bytes = new byte[length];
+        int bytesRead = read(bytes, 0, bytes.length);
+        checkArgument(bytesRead == length, "requested %s bytes but only %s was available", length, bytesRead);
+        return Slices.wrappedBuffer(bytes);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len)
+    {
+        Objects.checkFromIndexSize(off, len, b.length);
```

Comment:
We use `com.google.common.base.Preconditions.checkPositionIndexes` usually everywhere else